Tracks Opt-Out For Logged Out Users

leomattic opened the pull request:
We are improving the Tracks opt-out mechanism by making sure we do not record any Tracks events if the Analytics option has been disabled in the cookie banner.

Note the opt-out mechanism is already working for logged-in users.

Related to #79719 where we count the number of users opting out of analytics.

## Proposed Changes

* Within the function to record an event
    * Check for the `cookies sensitive_pixel_options`
    * If the Analytics option is turned off, we void recording the event

Something important here is that I am copy-pasting code from `client/lib/analytics/utils/` to `packages/calypso-analytics/src/utils/`. I would love to find a way to avoid the code duplication. Can we import files from there directly?
* `get-tracking-prefs.ts`
* `is-country-in-gdpr-zone.ts`
* `is-region-in-ccpa-zone.ts`


## Testing Instructions

* Run locally
* Make sure you see the cookie banner
* Unselect the Analytics option
* Confirm there are no more Tracks events with Tracks vigilante or watching network traffic

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

leomattic committed:
Do not record Tracks events if Analytics has been disabled via the cookie banner

leomattic committed:
add getTrackingPrefs

gmovr commented on config/development.json:
```diff
@@ -64,7 +64,7 @@
 		"fediverse/allow-opt-in": true,
 		"full-site-editing/beta-opt-in": true,
-		"cookie-banner": false,
+		"cookie-banner": true,
```
You can also use `?flags=cookie-banner` for testing. Or change these, but they shouldn't be committed to prod. :)

leomattic replied:
Thanks @gmovr ! I have changed back the option.
`?flags=cookie-banner` is a great tip! 
The setup has been a little confusing for me as I am not currently in a GDPR country and had to try different things to get the banner to show up 😅

robertsreberski replied:
> I am not currently in a GDPR country and had to try different things to get the banner to show up

The whole cookie banner depends on the `country_code` cookie - so you can easily switch countries by modifying it. You can use dev tools to change it to some European country e.g. `NL`.

gmovr commented on packages/calypso-analytics/src/utils/get-tracking-prefs.ts:
```diff
@@ -0,0 +1,102 @@
+export const TRACKING_PREFS_COOKIE_V1 = 'sensitive_pixel_option';
+export const TRACKING_PREFS_COOKIE_V2 = 'sensitive_pixel_options';
+
+export type TrackingPrefs = {
```
We have these types here `TrackingPrefs` client/lib/analytics/utils/get-tracking-prefs.ts, is this to move them to a package? If we do, could we also clean up the ones in client?

leomattic replied:
Hi hi @gmovr thanks so much for the review! I am very excited to help this move forward and improve our tracking a little :D 

One very big warning I should make is that I am new to javascript, calypso and to front-end development in general. This is my first PR 🥳 .

I am still getting the hang of components and packages. I think I am missing something obvious here. How can I use TrackingPrefs from here when it's in `client/lib/analytics/utils/` ? I agree we should keep the code in only one place.

Big thanks for the help!!

robertsreberski replied:
Hey @leomattic! I would be happy to help with that one. You indeed cannot import TrackingPrefs from `client/...`, but we could do it the other way around - it's just so as to not repeat the code.

You can replace uses of TrackingPrefs in `client/lib/analytics/utils/get-tracking-prefs.ts` and other files with `import { TrackingPrefs } from '@automattic/calypso-analytics'`.

gmovr reviewed (commented):
Haven't had time to fully review this, just leaving some notes. 👍

leomattic committed:
toggle back cookie banner in conf file

robertsreberski commented on packages/calypso-analytics/src/tracks.ts:
```diff
@@ -175,6 +176,18 @@ export function identifyUser( userData: any ): any {
 export function recordTracksEvent( eventName: string, eventProperties?: any ) {
 	eventProperties = eventProperties || {};
 
+	// We are using the same functions from client/lib/analytics
+	// Remember to check if we can avoid the code duplication
```
As mentioned, we can avoid that by importing the functions we moved to the `calypso-analytics` package in this PR in all use cases in `client/` code

MicBosi replied:
@leomattic I think we can just remove or update this comment right? Looks like now we don't have code duplication anymore?

leomattic replied:
Good catch @MicBosi !!
I am updating it now.

robertsreberski reviewed (changes requested):
The code looks good! It's great to see that we improve Tracks analytics to honour user tracking preferences.

The only thing left to do is to clean up the code duplication of methods:
- getTrackingPrefs
- parseTrackingPrefs
- isCountryInGdprZone
- isRegionInCcpaZone

and type `TrackingPrefs`.

As we copied them from `client/lib/analytics` we can remove them from there - and replace all occurrences with ones created in `calypso-analytics` (`import { TrackingPrefs } from '@automattic/calypso-analytics'`).

leomattic committed:
rebase

leomattic commented:
Hi @robertsreberski 👋  

I removed the files from `client/lib/analytics` and updated the references to `calypso-analytics` as suggested. I would love to get your help with a review as I am not totally sure how things work in javascript 🙏

leomattic committed:
update test using isRegionInCcpaZone

leomattic committed:
adjust behavour when not called on the client side

robertsreberski commented:
@leomattic the code looks good! thanks for making changes. one last thing to do is to prepare the package for publishing:

https://github.com/Automattic/wp-calypso/blob/trunk/docs/monorepo.md#publishing

it basically means to bump the version in package.json of the library code and to understand what actions to take after merging the PR.

MicBosi commented on packages/calypso-analytics/src/utils/get-tracking-prefs.ts:
```diff
@@ -60,13 +60,15 @@ export const parseTrackingPrefs = (
  *
- * WARNING: this function only works on the client side.
+ * WARNING: this function is meant to work on the client side. If not called
+ *          from the client side then it defaults to allo all
```
@leomattic Minor typo: "allo all" -> "allow all"?

leomattic replied:
Very true @MicBosi, thanks!

leomattic committed:
remove outdated comment

leomattic committed:
fix typo

MicBosi reviewed (approved):
I tested various scenarios, everything works as expected 🎉 

- The user preference gets picked up immediately as soon as the preferences cookie is changed or deleted and Tracks events are fired only when explicit consent is given while in GDPR countries
- Deleting the cookie immediately stops Tracks analytics in GDPR countries
- Changing country code cookie immediately resumes Tracks tracking regardless of preferences cookie settings

Awesome work @leomattic!!!

Many thanks @gmovr and @robertsreberski for helping us out!

@leomattic I guess next step for us is to carefully review and follow the steps in 

https://github.com/Automattic/wp-calypso/blob/trunk/docs/monorepo.md#publishing

that Robert mentioned.

MicBosi committed:
rebase

MicBosi commented:
I have updated the package version and changelog. I'm holding off merging atm in order to gather some data from the cookie banner view/ok events and share the results, as well as to make sure the deployment is coordinated with any test or experiment that might be going on.

I've also setup my npmjs account with the A8C organization in order to release the package right after merging.

travisw committed:
Trying to rebase or merge trunk was being a huge pain. This just manually applies the minimal changes to this branch 🤞

travisw committed:
fixing conflicts - undo a change to reapply later

travisw committed:
fixing conflicts - undo changes to reapply later

travisw committed:
reapply changes from before

travisw committed:
update changelog

MicBosi commented:
> I'm not sure if this was working correctly before the rebase or not though.

Actually, looks it was my fault to direct Leo to implement it so that Tracks events are blocked unless explicit consent is given, but I believe the correct behaviour is the one followed by the landing pages, ie, fire Tracks unless explicitly disabled.

I'll double check with legal just to have a definitive answer and to make sure this approach is still acceptable.

MicBosi commented:
@catscheer confirmed the LP behaviour is correct.

MicBosi committed:
Make Analytics tracking opt-out rather than opt-in

MicBosi commented:
@travisw I've committed the small change necessary to allow Analytics tracking to fire when in GDPR countries when the user hasn't expressed their preference yet - ie, you should be able to see Tracks events being fired while the cookie banner is visible, before clicking on it.

MicBosi committed:
Revert changes to see if TC tests work or not

MicBosi committed:
Make Analytics tracking opt-out rather than opt-in - Take II

leomattic commented:
Woot woot! 🍾
What an adventure!!

So happy to see this across the finish line.
Awesome job team!! 🥇

MicBosi commented:
1,000,000 thanks to you @leomattic and @travisw 👏 👏 👏
